Fix wrapsizer min size calculation

thesiv opened the pull request:
Fix the `wxWrapSizer::CalcMinFromMinor` function: at the end of the
calculation a `sumMinor` variable contain minor size sum without minor
size of a last line, so add `maxMinor` size to the result min size.

thesiv committed:
Fix wrapsizer min size calculation

Fix the `wxWrapSizer::CalcMinFromMinor` function: at the end of the
calculation a `sumMinor` variable contain minor size sum without minor
size of a last line, so add `maxMinor` size to the result min size.

vadz commented:
For the record: this needs unit tests, so will merging it until they're provided (even if the fix itself definitely seems correct).

thesiv committed:
Use CATCH in WrapSizer unit tests

Use CATCH instead of CppUnit for unit tests of wxWrapSizer.

thesiv committed:
Add unit test for wxWrapSizer::CalcMinFromMinor

Add a unit test for a special case of a wrap sizer min size caclulation.
Test wxWrapSizer::CalcMinFromMinor function for this case, when a wrap
sizer used inside a sizer with the same alignment.

vadz commented on tests/sizers/wrapsizer.cpp:
```diff
@@ -28,74 +28,85 @@
-
-// also include in its own registry so that these tests can be run alone
-CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( WrapSizerTestCase, "WrapSizerTestCase" );
+#define TEXT_VALIDATOR_TEST_CASE(name, tags) \
```
Assuming we really need this macro at all, it should have the right name:

```suggestion
#define WRAP_VALIDATOR_TEST_CASE(name, tags) \
```

vadz commented on tests/sizers/wrapsizer.cpp:
```diff
@@ -28,74 +28,85 @@
-    delete m_win;
-    m_win = NULL;
+    delete m_leftWin;
+    m_leftWin = NULL;
```
Setting members to null is pretty useless in the dtor...

vadz commented on tests/sizers/wrapsizer.cpp:
```diff
@@ -105,23 +116,54 @@ void WrapSizerTestCase::CalcMin()
+    const wxSize rightClientSize = m_rightWin->GetClientSize();
+
+    // Add three child windows. Sum of the first and the second windows widths should
+    // be less then the width of the third window.
```
```suggestion
    // be less than the width of the third window.
```

vadz commented on tests/sizers/wrapsizer.cpp:
```diff
@@ -28,74 +28,85 @@
 // ----------------------------------------------------------------------------
 
-class WrapSizerTestCase : public CppUnit::TestCase
+class WrapSizerTestCase
```
Sorry, I just don't understand what are you trying to do with this class. The idea of having a common test fixture like this is to reuse the same setup code for multiple tests, but here you seem to be using only `m_leftWin` in one test and only `m_rightWin` in the other one. What sense does it make to do it like this?

Either we should use the same `m_win` in both tests (as we did before for the single test) or maybe we shouldn't have any fixture class at all if there is just nothing to reuse... Or, finally, if the current approach is really useful, there should be a comment here explaining why is it so because this is not obvious at all from just looking at the code.

vadz commented on tests/sizers/wrapsizer.cpp:
```diff
@@ -105,23 +116,54 @@ void WrapSizerTestCase::CalcMin()
+
+TEXT_VALIDATOR_TEST_CASE("wxWrapSizer::CalcMinFromMinor", "[wxWrapSizer]")
+{
+    const wxSize rightClientSize = m_rightWin->GetClientSize();
```
As we create this window with a fixed, hardcoded client size, we could also use hardcoded sizes here IMO instead of `rightClientSize.x / 4 - 10` etc. This would only make the code more clear.

vadz commented on src/common/wrapsizer.cpp:
```diff
@@ -469,7 +469,7 @@ void wxWrapSizer::CalcMinFromMinor(int totMinor)
 
     // Now have min size in the opposite direction
-    m_calculatedMinSize = SizeFromMajorMinor(lineSize,sumMinor);
+    m_calculatedMinSize = SizeFromMajorMinor(lineSize, sumMinor+maxMinor);
```
After looking at the code in more details, I wonder if it wouldn't be better to update `sumMinor` inside the loop to account for the last (possibly incomplete) line? I think it would be more clear than just suddenly adding `maxMinor` here without any explanation (I know there is one in the commit message, but not in the code).

vadz reviewed (commented):
Sorry, the new test organization just doesn't make much sense to me. Could you please explain why did you do it like this or change it?

thesiv commented:
Fixed all issues.

vadz commented on tests/sizers/wrapsizer.cpp:
```diff
@@ -24,104 +24,96 @@
+    CHECK(sizeMinExpected == wrapSizer->CalcMin());
 
-    CPPUNIT_ASSERT_EQUAL( sizeMinExpected, m_sizer->CalcMin() );
+    delete win;
```
This should be handled using `wxScopedPtr` as windows remaining after the test end if something goes wrong in it screw up the subsequent tests which can be very annoying as it's usually not at all obvious that a failure in some completely unrelated test is due to this test failing and leaving its window on screen.

vadz reviewed (commented):
This is much better, thanks a lot!

Just one small issue remaining.

thesiv commented:
Fixed.

vadz commented:
Thanks! Will merge in a moment.
